gonvme_fc.go

```diff
@@ -0,0 +1,359 @@
+			// traddr:  nn-0x11aaa111a1111a11:aa-0x11aaa11111111a11
+			//
+			// =====Discovery Log Entry 1======
+			// trtype:  tcp
+			// adrfam:  ipv4
+			// subtype: nvme subsystem
+			// treq:    not specified
+			// portid:  2304
+			// trsvcid: 4420
+			// subnqn:  nqn.1111-11.com.dell:powerstore:00:a1a1a1a111a1111a111a
+			// traddr:  1.1.1.1
+			// sectype: none
+
+			tokens := strings.Fields(line)
+			if len(tokens) < 2 {
+				continue
+			}
+			key := tokens[0]
+			value := strings.Join(tokens[1:], "")
+			switch key {
+
+			case "=====Discovery":
+				// add to array
+				if entryCount != 0 && !skipIteration && nvmeTarget.Portal == targetAddress {
+					targets = append(targets, nvmeTarget)
+				}
+				nvmeTarget = NVMeTarget{}
+				nvmeTarget.HostAdr = initiatorAddress
+				skipIteration = false
+				entryCount++
+				continue
+
+			case "trtype:":
+				nvmeTarget.TargetType = value
+				if value != NVMeTransportTypeFC {
+					skipIteration = true
+				}
+				break
+
+			case "traddr:":
+				nvmeTarget.Portal = value
+				break
+
+			case "subnqn:":
+				nvmeTarget.TargetNqn = value
+				break
+
+			case "adrfam:":
+				nvmeTarget.AdrFam = value
+				break
+
+			case "subtype:":
+				nvmeTarget.SubType = value
+				break
+
+			case "treq:":
+				nvmeTarget.Treq = value
+				break
+
+			case "portid:":
+				nvmeTarget.PortID = value
+				break
+
+			case "trsvcid:":
+				nvmeTarget.TrsvcID = value
+				break
+
+			case "sectype:":
+				nvmeTarget.SecType = value
+				break
+
+			}
+		}
+		if !skipIteration && nvmeTarget.TargetNqn != "" && nvmeTarget.Portal == targetAddress {
+			targets = append(targets, nvmeTarget)
+		}
+	}
+
+	if len(targets) == 0 {
+		fmt.Printf("\n Error discovering NVMe/FC targets: %s", err)
+		return []NVMeTarget{}, err
+	}
+
+	// TODO: Add optional login
+	// log into the target if asked
+	/*if login {
+		for _, t := range targets {
+			gonvme.PerformLogin(t)
+		}
+	}*/
+
+	return targets, nil
+}
+
+// NVMeConnect will attempt to connect into a given nvme target
+func (nvme *NVMeFC) NVMeConnect(target NVMeTarget) error {
+	return nvme.nvmeConnect(target)
+}
+
+func (nvme *NVMeFC) nvmeConnect(target NVMeTarget) error {
+	// nvme connect is done via the nvme cli
+	// nvme connect -t fc -a traddr -w host_traddr -n target_nqn
+	// where traddr = nn-<Target_WWNN>:pn-<Target_WWPN> and host_traddr = nn-<Initiator_WWNN>:pn-<Initiator_WWPN>
+	exe := nvme.buildNVMeCommand([]string{NVMeCommand, "connect", "-t", "fc", "-a", target.Portal, "-w", target.HostAdr, "-n", target.TargetNqn})
+	cmd := exec.Command(exe[0], exe[1:]...)
+	_, err := cmd.Output()
+
+	if err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			// nvme connect exited with an exit code != 0
+			nvmeConnectResult := -1
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				nvmeConnectResult = status.ExitStatus()
+			}
+			if nvmeConnectResult == 114 || nvmeConnectResult == 70 {
```

Comment:
are these error codes specific to any OS/platform? both codes denote session already exists?

Reply:
Yes, the error codes are different for nvme connect in different platforms